include/exadg/solvers_and_preconditioners/preconditioners/preconditioner_amg.h

```diff
@@ -154,17 +274,201 @@ class PreconditionerML : public PreconditionerBase<double>
     // re-calculate matrix
     pde_operator.calculate_system_matrix(system_matrix);
 
-    // initialize Trilinos' AMG
-    amg.initialize(system_matrix, ml_data);
+    if(operator_type == AMGOperatorType::Default)
+    {
+      // Default setup.
+      amg.initialize(system_matrix, ml_data);
+    }
+    else
+    {
+      // get Teuchos::ParameterList to provide custom near null space basis
+      Teuchos::ParameterList parameter_list = get_parameter_list();
+
+      std::vector<VectorType> constant_modes = get_constant_modes();
```

Comment:
Is there a reason why this function is called outside `set_operator_nullspace()`? Does the comment below "has to stay alive" also hold for this? I think the code would be easier to understand when integrating this into `set_operator_nullspace()`, because dependencies are not clear in the present code in my opinion.

Reply:
No I just thought that its easier to see what happens if I block the code into some parts that do only a limited thing instead of having 1 blob of code.
Now it is outside since we use deal.II functionality :)